ta/trusted_key/entry.c

```diff
@@ -0,0 +1,308 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2019-20, Linaro Limited
+ */
+
+#include <pta_system.h>
+#include <string.h>
+#include <string_ext.h>
+#include <tee_internal_api.h>
+#include <tee_internal_api_extensions.h>
+#include <trusted_key.h>
+#include <util.h>
+
+#define IV_SIZE			16
+#define TAG_SIZE		16
+#define MAX_BUF_SIZE		512
+
+static TEE_Result get_random(uint32_t types,
+			     TEE_Param params[TEE_NUM_PARAMS])
+{
+	FMSG("Invoked TA_CMD_GET_RANDOM");
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_OUTPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (!params[0].memref.buffer || params[0].memref.size == 0)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	TEE_GenerateRandom(params[0].memref.buffer, params[0].memref.size);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result derive_unique_key(uint8_t *key, uint16_t key_size,
+				    uint8_t *extra, uint16_t extra_size)
+{
+	TEE_TASessionHandle sess = TEE_HANDLE_NULL;
+	TEE_Param params[TEE_NUM_PARAMS] = { 0 };
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint32_t ret_orig = 0;
+	uint32_t param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+					       TEE_PARAM_TYPE_MEMREF_OUTPUT,
+					       TEE_PARAM_TYPE_NONE,
+					       TEE_PARAM_TYPE_NONE);
+
+	res = TEE_OpenTASession(&(const TEE_UUID)PTA_SYSTEM_UUID,
+				TEE_TIMEOUT_INFINITE, 0, NULL, &sess,
+				&ret_orig);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	if (extra && extra_size > 0) {
+		params[0].memref.buffer = extra;
+		params[0].memref.size = extra_size;
+	}
+
+	params[1].memref.buffer = key;
+	params[1].memref.size = key_size;
+
+	res = TEE_InvokeTACommand(sess, TEE_TIMEOUT_INFINITE,
+				  PTA_SYSTEM_DERIVE_TA_UNIQUE_KEY,
+				  param_types, params, &ret_orig);
+
+	TEE_CloseTASession(sess);
+
+	return res;
+}
+
+static TEE_Result huk_AE_encrypt(TEE_OperationHandle crypto_op, void *in,
+				 size_t in_sz, void *out, size_t *out_sz)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint8_t *iv = (uint8_t *)out;
+	uint8_t *tag = (uint8_t *)out + IV_SIZE;
+	uint8_t *enc_key = (uint8_t *)out + IV_SIZE + TAG_SIZE;
```

Comment:
Useless casts

Reply:
Need to keep such casts otherwise compiler reports following warning:

ta/trusted_keys/entry.c:115:24: warning: pointer of type ‘void *’ used in arithmetic [-Wpointer-arith]

Reply:
Do the parameter have to be void pointers? Can't they be uint8_t pointers? I.e., make the cast in {un}seal_trusted_key directly?

Reply:
Okay.

Reply:
@b49020 my mistake, indeed only the first cast is superfluous. That said, @jbech-linaro +1